tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -402,6 +413,329 @@ impl ServerKey {
+            to the number of blocks of the right hand side: lhs {} blocks, rhs {} blocks",
+            lhs.len(),
+            rhs.len()
+        );
+
+        if lhs.is_empty() {
+            // Then both are empty
+            if compute_overflow {
+                return Some(self.create_trivial_boolean_block(false));
+            }
+            return None;
+        }
+
+        for (lhs_b, rhs_b) in lhs.iter_mut().zip(rhs.iter()) {
+            self.key.unchecked_sub_assign(lhs_b, rhs_b);
+        }
+        if let Some(borrow) = input_borrow {
+            self.key.unchecked_sub_assign(&mut lhs[0], &borrow.0);
+        }
+
+        // First step
+        let (shifted_blocks, mut block_states) =
+            self.compute_shifted_blocks_and_block_borrow_states(lhs);
+
+        // The propagation state of the last block will be used to determine
+        // if overflow occurs (i.e is there an output borrow)
+        let mut overflow_block = block_states.pop().unwrap();
+
+        let block_modulus = self.message_modulus().0 * self.carry_modulus().0;
+        let num_bits_in_block = block_modulus.ilog2();
+
+        // Just in case we compare with max noise level, but it should always be num_bits_in_blocks
+        // with the parameters we provide
+        let grouping_size = (num_bits_in_block as usize).min(self.key.max_noise_level.get());
+
+        // Second step
+        let (mut prepared_blocks, resolved_borrows) = {
+            let (propagation_simulators, resolved_borrows) = self
+                .compute_propagation_simulators_and_groups_carries(grouping_size, &block_states);
+
+            let mut prepared_blocks = shifted_blocks;
+            prepared_blocks
+                .iter_mut()
+                .zip(propagation_simulators.iter())
+                .for_each(|(block, simulator)| {
+                    // simulator may have either of these value
+                    // '2' if the block is borrowed from
+                    // '1' if the block will be borrowed from if the group it belongs to receive a
+                    //     borrow
+                    // '0' if the block will absorb any potential borrow
+                    //
+                    // What we do is we subtract this value from the block, as it's a borrow, not a
+                    // carry, and we add one, this means:
+                    //
+                    // '- (2 + 1) ==  -1' We remove one if the block was meant to receive a borrow
+                    // '- (1 + 1) ==  -0' The block won't change, which means that when subtracting
+                    // the borrow (value: 1 or 0) that the group receives, its correctly applied
+                    // i.e the propagation simulation will be correctly done
+                    // '- (0 + 1) ==  +1' we add one, meaning that if the block receives a borrow,
+                    // we would remove one from the block, which would be absorbed by the 1 we just
+                    // added
```

Comment:
not super clear here on what that means

Reply:
Whe have a shifted block that looks like `0xx0`, to put the 'propagation simulator block', the block that simulates how the input borrow of the group is propagated (or not) to the block in question, we don't do `0xx0 + 00pp`, we do `0xx0 - 00pp + 0001`.

So if `00pp == 0000` (meanins any borrow is the group receives is actually absorbed an so it does not reach the block
This translates in us doing o `0xx0 - 0000 + 0001 = 0xx1`, so if I the group was to be borrowed from, we need to sub 1 from it, `0xx1 - 0001 = 0xx0` , the borrow has been absorbed, and the block is untouched by it as it should have been